feat: staking token liens to prevent transfer

JimLarson opened the pull request:
See #3277

Implements lien functionality by reusing locking (aka vesting account) mechanisms.

Compared to earlier versions of the design, statefully holds total liened amount in cosmos-sdk storage to prevent need for synchronous upcalls to Swingset.

Still TBD:

* Tests for LIEN_SET_TOTAL messages.
* The Go-to-JS LIEN_SLASHED message.
* JS middleware.
* Attestation contract updates to use the new API

michaelfig commented on golang/cosmos/x/lien/lien.go:
```diff
@@ -0,0 +1,140 @@
+
+// msgAccountState marshals into the AccountState message for the lien bridge.
+type msgAccountState struct {
+	CurrentTime string `json:"currentTime"`
```
The `CurrentTime` should be an `int64`, whatever its interpretation is.

JimLarson replied:
Done - assuming UNIX epoch.

michaelfig commented on golang/cosmos/x/lien/lien.go:
```diff
@@ -0,0 +1,140 @@
+	Type    string `json:"type"`
+	Address string `json:"address"`
+	Denom   string `json:"denom"`
+	Amount  string `json:"amount"`
```
```suggestion
	Amount  sdk.Int `json:"amount"`
```

JimLarson replied:
Done.

michaelfig commented on golang/cosmos/x/lien/lien.go:
```diff
@@ -0,0 +1,140 @@
+// msgAccountState marshals into the AccountState message for the lien bridge.
+type msgAccountState struct {
+	CurrentTime string `json:"currentTime"`
+	Total       string `json:"total"`
+	Bonded      string `json:"bonded"`
+	Unbonding   string `json:"unbonding"`
+	Locked      string `json:"locked"`
+	Liened      string `json:"liened"`
```
All these should be `sdk.Int`s.

JimLarson replied:
Done.

michaelfig commented on golang/cosmos/x/lien/lien.go:
```diff
@@ -0,0 +1,140 @@
+	}
+	state := ch.keeper.GetAccountState(ctx, addr)
+	reply := msgAccountState{
+		CurrentTime: ctx.BlockTime().String(), // REVIEWER: what format?
```
I believe `ctx.BlockTime()` is an int64 of seconds since the epoch.

JimLarson replied:
It's a `time.Time`, but we'll represent it as an `int64` of seconds since UNIX epoch.

michaelfig commented on golang/cosmos/x/lien/lien_test.go:
```diff
@@ -0,0 +1,83 @@
+	if err != nil {
+		t.Fatalf("cannot unmarshal reply %s: %v", reply, err)
+	}
+	acctState.CurrentTime = "" // TODO: make predictable
+	want := msgAccountState{
+		CurrentTime: "", Total: "123", Bonded: "0", Unbonding: "0", Locked: "0", Liened: "0",
```
I'd suggest:
```suggestion
	want := msgAccountState{
		CurrentTime: acctState.CurrentTime, Total: "123", Bonded: "0", Unbonding: "0", Locked: "0", Liened: "0",
```

JimLarson replied:
Done.

michaelfig commented on golang/cosmos/x/lien/spec/01_concepts.md:
```diff
@@ -0,0 +1,129 @@
+## Notes on Terminology
+
+* "Bonded", "delegated", and "staked" are all synonyms.
+* "Unlocking" is a time-delayted process by which tokens become available for
```
```suggestion
* "Unlocking" is a time-delayed process by which tokens become available for
```

JimLarson replied:
Done.

michaelfig commented on golang/cosmos/x/lien/spec/01_concepts.md:
```diff
@@ -0,0 +1,129 @@
+unbonded when a validator retires. The best we can do is that when
+the higher layers wish to increase the total liened amount, the new
+total must be less than the bonded tokens and less than the unlocked
+tokens.  Liens can be lifted at any time, 
```
Maybe this is the end of the sentence?

```suggestion
tokens.  Liens can be lifted at any time.
```

JimLarson replied:
Done.

michaelfig commented on golang/cosmos/x/lien/spec/01_concepts.md:
```diff
@@ -0,0 +1,129 @@
+
+## Implementing Lien Encumbrance
+
+Unliened tokens may not be transferred out of an account. Rather than
```
```suggestion
Liened tokens may not be transferred out of an account. Rather than
```

JimLarson replied:
Done. That makes more sense.

michaelfig commented on golang/cosmos/x/lien/types/accountkeeper.go:
```diff
@@ -0,0 +1,81 @@
+	Unwrap: identWrap,
+}
+
+// WrappedAccountKeeper wraps an account AccountKeeper insert an outer wrapper
```
```suggestion
// WrappedAccountKeeper wraps an account AccountKeeper
```

JimLarson replied:
Fixed with different wording.

michaelfig commented on golang/cosmos/x/lien/types/accountkeeper.go:
```diff
@@ -0,0 +1,81 @@
+	wak.AccountKeeper.SetAccount(ctx, unwrappedAcc)
+}
+
+// SetWrapper updates the AcountWrapper.
```
```suggestion
// SetWrapper updates the AccountWrapper.
```

JimLarson replied:
Done.

michaelfig commented on golang/cosmos/x/lien/types/key.go:
```diff
@@ -0,0 +1,29 @@
+	StoreKey = ModuleName
+)
+
+var (
+	LienByAddressPrefix = []byte{0x01}
+)
```
I don't think you need a prefix in this case.  The only reason the "storage" uses it is because the root node's key is an just an empty `[]byte{}`, which cannot be represented in the KVStore.  So, I hacked around their hack by adding a prefix to all the keys.

JimLarson replied:
Huh, I thought it was to preserve the ability to add different data in the same store in the future. In fact, look at all the different prefixes used in `cosmos-sdk/x/staking/types/keys.go`. I'd rather keep the prefix if it's not hurting anything.

michaelfig replied:
A key prefix is a hack that we should avoid if possible.  If you want to add different data in the same store, there's a better pattern:

```
import (
        "github.com/cosmos/cosmos-sdk/store/prefix"
)

var (
         subPrefix = []byte("some-kinda-prefix")
)

// ... and in a keeper function:
        store := ctx.KVStore(k.storeKey)
        subStore := prefix.NewStore(store, subPrefix)
```

JimLarson replied:
Groovy. Leaving the `address <--> key` conversion API, but no prefix.

michaelfig commented on packages/cosmic-swingset/src/block-manager.js:
```diff
@@ -10,6 +10,7 @@ const DELIVER_INBOUND = 'DELIVER_INBOUND';
 const END_BLOCK = 'END_BLOCK';
 const COMMIT_BLOCK = 'COMMIT_BLOCK';
 const IBC_EVENT = 'IBC_EVENT';
+const LIEN_GET_LIENED_AMOUNT = 'LIEN_GET_LIENED_AMOUNT';
```
I think we've removed this upcall now.

JimLarson replied:
Reverted.

michaelfig commented on packages/cosmic-swingset/src/block-manager.js:
```diff
@@ -112,6 +113,11 @@ export default function makeBlockManager({
         break;
       }
 
+      case LIEN_GET_LIENED_AMOUNT: {
+        const lockedAmount = 0n;
+        return await Promise.resolve(JSON.stringify(String(lockedAmount)));
+      }
+
```
Likewise, please remove this clause.

JimLarson replied:
Reverted.

michaelfig reviewed (commented):
This is really good progress!

I'd like to suggest splitting out an `x/vlien` module from `x/lien`.  The distinction between the two is that `vlien` would include the JSON structures, upcalls to a "controller", and downcalls to the "port".  That part of the PR would likely never be merged into Cosmos SDK.  The `lien` part, instead, would be pure Golang and provide more "Go-ish" APIs (such as a hooks mechanism) that don't involve the "controller" per se.

But as before, let's focus on the priorities first and come back to this in a little bit.

JimLarson commented on golang/cosmos/proto/agoric/lien/genesis.proto:
```diff
@@ -0,0 +1,27 @@
+// The lien on a particular account
+message AccountLien {
+    // Account address, bech32-encoded.
+    string address = 1;
```
A quick survey of the cosmos-sdk proto files shows only raw strings like this, and the verification is simply `sdk.AccAddressFromBech32()`. If you see a better way, please refer me to the location, as I don't see anything on first pass.

JimLarson commented on golang/cosmos/x/lien/keeper/account.go:
```diff
@@ -0,0 +1,169 @@
+	indexA, indexB := 0, 0
+	for indexA < len(a) && indexB < len(b) {
+		coinA, coinB := a[indexA], b[indexB]
+		switch strings.Compare(coinA.Denom, coinB.Denom) {
```
Done.

JimLarson commented on golang/cosmos/x/lien/keeper/account_test.go:
```diff
@@ -0,0 +1,111 @@
+		t.Fatalf("unwrapper did not produce a base account: %+v", unwrapped)
+	}
+	if baseAccount.AccountNumber != 17 {
+		t.Errorf("wong account number %d, want 17", baseAccount.AccountNumber)
```
Done.

JimLarson commented on golang/cosmos/x/lien/keeper/keeper.go:
```diff
@@ -0,0 +1,192 @@
+
+// stakingToken is the single denomination used for staking.
+// The lien API pays lip service to the idea that staking is done with
+// arbitrary sdk.Coins, but the UnboindingDelegation contains only a bare
```
Done.

JimLarson commented:
I've pushed another commit to address the review comments. I'll need to rebase to update and fix the build, but doing so will screw up your current comments, so please review to just confirm that I've satisfactorily addressed your comments, then I'll update.

JimLarson commented:
PTAL

michaelfig commented on golang/cosmos/x/lien/lien.go:
```diff
@@ -0,0 +1,130 @@
+		ch.keeper.SetLien(ctx, addr, lien)
+		return "true", nil
+	} else {
+		// check if it's okay to increase lein
```
Please only change this if you're forced to rebase.
```suggestion
		// check if it's okay to increase lien
```

michaelfig reviewed (approved):
LGTM!  Well done!!
